### PR TITLE
Add S2A Half Connection

### PR DESCRIPTION
**Pull request by Ryanfsdf**
Added the initial S2A Half Connection. This is largely unimplemented, but exposes the basic methods that S2A Half Connection would provide. This also implements `expandLabel` specified in https://tools.ietf.org/html/rfc8446#section-7.1. It basically constructs a HkdfLabel struct (Which is a struct specific to TLS 1.3) and passes that to the underlying HKDF expand function.

One thing that became apparent when implementing `expandLabel` was that calling `h().Size()` is expensive. For example, `h` that is passed in will most likely be something like `sha256.New`. When we call `h()`, we are calling `sha256.New()` and creating a new Hash just to get the length. I think this is quite expensive when all we want is the length of the hash. We are doing this in both `expandLabel` as well as the underlying `expand` function. I think it may actually be worthwhile to bring back the `length` field in the expand interface for the underlying HKDF expander since we call `hash().Size()` here already. Either that, or just create a `hashLength` constant somewhere and hardcode the length, and then specify in the function documentation that the hashing function *must* return a hash of a certain length. WDYT?

Also, I'm not sure if `S2AHalfConnection` will be used outside of this package. I've exported it for now. Please let me know if it needs to be unexported.

---
This change is [Reviewable](https://reviewable.io/reviews/matthewstevenson88/grpc-go/14)

### CODE REVIEW

File: security/s2a/internal/crypter/halfconn.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	"hash"
+)
+
+type S2AHalfConnection struct {
+	aeadCrypter s2aAeadCrypter
+	expander    hkdfExpander
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:11]: The half connection will not need to own an expander. It will just create one when it is needed (which will only be at creation time and when a key update occurs).

File: security/s2a/internal/crypter/halfconn.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+	sequence    counter
+}
+
+func NewHalfConn(aeadCrypter s2aAeadCrypter, expander hkdfExpander) S2AHalfConnection {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:15]: The `NewHalfConn` API should accept 2 arguments: a "ciphersuite" and a "traffic secret".

The "ciphersuite" could be an enum or a string or similar, it's up to you. The options should be AES-128-GCM-SHA256, AES-256-GCM-SHA284, and CHACHA20-POLY1305-SHA256.

The "traffic secret" should be a `byte[]` of a particular length that depends on the "ciphersuite". See [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L219) for how this should work.

The implementation of `NewHalfConn` should do the following:

1. Save the traffic secret as a member field.
2. Use the traffic secret and HKDF-expand to derive the key; see [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L147).
3. Use the traffic secret and HKDF-expand to derive the nonce; see [the C++ version](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L159).
4. Use the key to create an AEAD crypter.
5. Save the nonce as a member field.

File: security/s2a/internal/crypter/halfconn.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+func (hc *S2AHalfConnection) UpdateKey(key []byte) error {
+	// TODO(rnkim): Implement this.
+	panic("UpdateKey currently unimplemented")
+}
+
+// expandLabel implements HKDF-Expand-Label specified in
+// https://tools.ietf.org/html/rfc8446#section-7.1. Note that the `Context` and
+// `Length` parameters have been omitted since we always pass in an empty string
+// for `Context` and we used a fixed length.
+func (hc *S2AHalfConnection) expandLabel(h func() hash.Hash, secret, label []byte) ([]byte, error) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:38]: I think we can model this function off the [C++ derive_secret API](https://github.com/matthewstevenson88/grpc/blob/b4521fa8f145bd319d91d98fcaaed7ad7018d67e/src/core/tsi/s2a/record_protocol/s2a_crypter.cc#L119). To do so, we would not add "tls13 " to `label`. Rather, the caller will plug in values like "tls13 key" or "tls13 upd" for `label` when the caller wants to derive a key, or do a key update.

**Comment by matthewstevenson88** [comment]
Thanks for the PR Ryan! I've left some general comments. PTAL.

**Comment by matthewstevenson88** [comment]
Also I missed your comment about bringing back the length function to the expander: that is totally fine by me! Can you send a separate PR to do so with the justification?

File: security/s2a/internal/crypter/halfconn.go (hunk 3 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:38]: Was there any specific reason the C++ version was modeled this way? I believe "tls13 " is always a prefix for the HkdfLabel specified in the TLS 1.3 RFC. It seems unnecessary to prepend "tls13 " in front of every label when it will never change.

File: security/s2a/internal/crypter/halfconn.go (hunk 1 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:11]: Do you think it may be worth keeping as a member still? I'm assuming the half connection may have its key updated several times, in which case we would create a new expander every time. If we keep it as a member, I believe we can just reuse the same expander instead of creating one every time.

Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:11]: If you want to keep it, that's totally fine so long as `expander` doesn't hold any state. If that's the case, then we might as well keep it around.

File: security/s2a/internal/crypter/halfconn.go (hunk 3 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:38]: For C++, the reason was just that it was simpler - in S2A-Go terminology, the choices were:

1. There is  `expandLabel` (implemented using HKDF-expand), called by `deriveSecret`, which is called by `deriveKey`, `deriveNonce`, and `advanceTrafficSecret`.
2. There is `deriveSecret` (implemented using HKDF-expand), which is called by `deriveKey`, `deriveNonce`, and `advanceTrafficSecret`.

Option 2 was used because the `expandLabel` API wasn't needed: we had to put in a different label for each of the 3 uses, so might as well put in the label prefix as well.

File: security/s2a/internal/crypter/halfconn.go (hunk 1 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:11]: The struct is empty `type defaultHKDFExpander struct{}` so it doesn't hold any state at all. I think it also might be worth keeping because I think the code would look better if we call `hc.expander.expand(...)` in various places rather than `newDefaultHKDFExpander().expand(...)`.

If we have to call `newDefaultHKDFExpander().expand(...)` everywhere, I think that would defeat the purpose of creating an interface for the expansion function to begin with since we would have to swap all of that to `new<Different_Library>HKDFExpander().expand(...)` in multiple places.

File: security/s2a/internal/crypter/halfconn.go (hunk 3 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:38]: Thanks for the explanation! Do you think it would be worth passing in the entire label with the "tl13 " prepended to it like in C++? I think it might be worth saving a couple of bytes here by not passing it to every call of `expandLabel` or `deriveSecret`, but there's also some value in keeping it consistent with the C++ version as well.

File: security/s2a/internal/crypter/halfconn.go (hunk 2 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:15]: Thanks for explaining and linking the C++ version as reference.

In terms of the `ciphersuite` parameter, do you think it might be worth creating a `ciphersuite` interface? With 3 methods: `KeySize`, `NonceSize`, `ExpectedTrafficSecretSize`. This way, we won't need to use an enum + switch statement.

Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:15]: I realized we should just use the `Ciphersuite` defined in the proto file, since that is the easiest thing for the caller to use. This is defined in `common.pb.go`, which I've added in #16 . Sorry for not thinking of this before!

File: security/s2a/internal/crypter/halfconn.go (hunk 3 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:38]: I think there's probably no real performance difference, unless adding a call to the stack trace is more expensive in Go than I've been led to believe. :)

I leave it up to you. Personally, I think removing the `expandLabel` and only having `deriveSecret` (that takes in the full label) is easier for a reader to understand. But I don't have very strong feelings either way.

File: security/s2a/internal/crypter/ciphersuite.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+)
+
+type ciphersuite interface {
+	// keySize returns the key size in bytes.
+	keySize() int
+	// nonceSize returns the nonce size in bytes.
+	nonceSize() int
+	// expectedTrafficSecretSize returns the expected traffic secret size in
+	// bytes.
+	expectedTrafficSecretSize() int
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/ciphersuite.go:17]: Can we call this `trafficSecretSize`? This would keep the naming in line with the other functions.

File: security/s2a/internal/crypter/ciphersuite.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
+package crypter
+
+import (
+	"crypto/sha256"
+	"golang.org/x/crypto/sha3"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal"
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/ciphersuite.go:6]: Is this importing everything in `internal`? I think you only need `common.pb.go`.

File: security/s2a/internal/crypter/halfconn.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	expander      hkdfExpander
+	seqCounter    counter
+	trafficSecret []byte
+	nonce         []byte
+}
+
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := NewCiphersuite(ciphersuite)
+	if cs.expectedTrafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bits, given: %v", cs.expectedTrafficSecretSize()*8, trafficSecret)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:21]: Please make all references to bytes rather than bits.

File: security/s2a/internal/crypter/halfconn.go [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+	}
+
+	hc := S2AHalfConnection{expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
+	key, err := hc.deriveSecret(cs.hashFunction(), trafficSecret, []byte("tls13 key"))
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:25]: Please make "tls13 key" and "tls13 iv" constants.

File: security/s2a/internal/crypter/halfconn.go [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+	panic("Decrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) UpdateKey(key []byte) error {
+	// TODO(rnkim): Implement this.
+	panic("UpdateKey currently unimplemented")
+}
+
+// deriveSecret implements Derive-Secret specified in
+// https://tools.ietf.org/html/rfc8446#section-7.1. Note that the `Context` and
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:56]: I think we can omit the discussion of Context and Length from here, since they are notions from the RFC. If the reader wants, they can map the RFC's function to yours using the link you gave.

File: security/s2a/internal/crypter/halfconn.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+	// TODO(rnkim): Implement this.
+	panic("Encrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) Decrypt(dst, ciphertext, aad []byte) ([]byte, error) {
+	// TODO(rnkim): Implement this.
+	panic("Decrypt currently unimplemented")
+}
+
+func (hc *S2AHalfConnection) UpdateKey(key []byte) error {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:50]: This function should not have a parameter. It has to derive the new key from the traffic secret.

File: security/s2a/internal/crypter/halfconn.go (hunk 8 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:56]: The original reason I referenced it here is because we modified the interface for "Expand-Label" specified in the RFC. However, since we updated the function name to `deriveSecret` rather than `expandLabel`, I'll remove this part of the comment since "Derive-Secret" doesn't have "Context" and "Length" as parameters anyways.

File: security/s2a/internal/crypter/ciphersuite.go (hunk 5 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/ciphersuite.go:6]: I think importing `internal` is as granular as it gets since we have to import entire packages in Go. However, I noticed that you merged a PR that moved the `common.pb.go` file over. Once I rebase, I assume I won't have to import `internal` anymore.

File: security/s2a/internal/crypter/ciphersuite.go (hunk 4 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/ciphersuite.go:17]: I called it `expectedTrafficSecretSize` because that is what it is called in the C version. I thought there might be a reason why it had the `expected` prepended to it whereas the other variables don't. I think just `trafficSecretSize` is more fitting though, so I've changed it.

File: security/s2a/internal/crypter/common.go [hunk 10]
```diff
@@ -7,6 +7,12 @@ const (
 	// nonceSize is the size of the nonce in number of bytes for
 	// AES-128-GCM-SHA256, AES-256-GCM-SHA384, and CHACHA20-POLY1305-SHA256.
 	nonceSize = 12
+	// sha256DigestLength is the digest length of sha256 in bytes.
+	sha256DigestLength = 32
+	// sha384DigestLength is the digest length of sha384 in bytes.
+	sha384DigestLength = 48
+	// uint64Size is the size of a uint64 in bytes.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/common.go:14]: Do we need this? Isn't this the definition of a `uint64`?

File: security/s2a/internal/crypter/halfconn.go [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+	nonce         []byte
+}
+
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := NewCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
+	}
+
+	hc := S2AHalfConnection{h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter()}
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:33]: Can we move this line after we derive the key and the nonce? That way, if an error occurs while deriving the key/nonce then we don't create the half connection at all.

File: security/s2a/internal/crypter/halfconn.go [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.decrypt(dst, ciphertext, nonce, aad)
+}
+
+func (hc *S2AHalfConnection) UpdateKey() error {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:75]: The entire body of this function should be under the lock (to guard against concurrent calls to `Encrypt` and `UpdateKey`).

File: security/s2a/internal/crypter/halfconn.go [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+	if err != nil {
+		return 0, err
+	}
+	hc.seqCounter.inc()
+	return sequence, nil
+}
+
+func (hc *S2AHalfConnection) maskedNonce(sequence uint64) []byte {
+	for i := 0; i < uint64Size; i++ {
+		hc.nonce[nonceSize-8+i] ^= sequence >> (56 - 8*i)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:108]: Could you add an extra set of parentheses around `sequence >> (56-8*i)`? Doing so will make this a bit more readable, I think.

Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:108]: The linter complains about unnecessary parenthesis so I think it's best that we follow what the linter suggests.

File: security/s2a/internal/crypter/halfconn.go [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 upd"
+)
+
+type S2AHalfConnection struct {
+	h             func() hash.Hash
+	aeadCrypter   s2aAeadCrypter
+	expander      hkdfExpander
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:20]: Could we

File: security/s2a/internal/crypter/ciphersuite.go (hunk 5 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/ciphersuite.go:6]: SGTM!

File: security/s2a/internal/crypter/halfconn.go (hunk 11 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:33]: I think you mean "then we don't create an AEAD crypter at all" correct? Since we don't make a half connection at all if there's a failure anywhere at all.

File: security/s2a/internal/crypter/halfconn.go (hunk 13 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:108]: Ah ok. Let's follow the linter then. :)

File: security/s2a/internal/crypter/common.go (hunk 10 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/common.go:14]: We need this otherwise we'll have to use `8` as an arbitrary constant everywhere rather than `uint64Size`. We use this in the nonce mask function.

File: security/s2a/internal/crypter/halfconn.go (hunk 11 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/halfconn.go:33]: But doesn't the `hc := S2AHalfConnection(...)` line initialize the half connection? (Modulo some variables not being set.)

In any event, I think it's best to fail fast, i.e. derive the key/nonce sooner (since those operations can fail) and then initialize `hc`.

File: security/s2a/internal/crypter/common.go (hunk 10 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/common.go:14]: But don't we already have a bunch of `8`'s in the `maskNonce` function anyway? And that seems to be the only place that `uint64Size` is used, unless I'm missing something.

If you want, you can add a comment in the `maskNonce` implementation that describes where the `8`'s are coming from?

**Comment by matthewstevenson88** [comment]
Thanks Ryan! The PR is looking really good - I left a few nits. Additionally, can you please add comments to the half connection's functions and tests for the half connection?

File: security/s2a/internal/crypter/halfconn.go (hunk 11 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:33]: Oh I see what you're saying, I think I misunderstood your first comment. `hc := S2AHalfConnection(...)` initializes the half connection but an empty half connection is returned on any error. We still need the half connection before we call `deriveSecret` since `deriveSecret` is a method of the half connection. I've removed the hash function from the parameter of `deriveSecret` just now since it has access to the hash function from the half connection anyways.

File: security/s2a/internal/crypter/common.go (hunk 10 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/common.go:14]: Oh those 8's were unintentional, those were meant to be replaced with `uint64Size`. I'll just remove the constant then and add a comment.

File: security/s2a/internal/crypter/halfconn.go (hunk 11 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/halfconn.go:33]: If we want to move `hc := S2AHalfConnection(...)` after we derive the secrets, we'll have to include the hash function as well as the expander in the `deriveSecret` API. I think it would be cleaner just to initialize the half connection first and then call `deriveSecret` after. The cost to initializing the half connection is minimal anyways, and it makes the code cleaner in my opinion (due to the cleaner `deriveSecret` parameters). WDYT?

File: security/s2a/internal/crypter/common.go (hunk 10 above)
Comment (Ryanfsdf) [reply security/s2a/internal/crypter/common.go]: On second thought, I just moved the constant over to the half connection file. I think having the constant is still useful.

File: security/s2a/internal/crypter/halfconn.go [hunk 15]
```diff
@@ -0,0 +1,165 @@
...
+
+const (
+	tls13Key    = "tls13 key"
+	tls13Nonce  = "tls13 iv"
+	tls13Update = "tls13 traffic upd"
+)
+
+type S2AHalfConnection struct {
+	cs            ciphersuite
+	h             func() hash.Hash
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:22]: Would it make sense to pass the hash function to the expander, and the expander can own the hash function? Unless I'm missing something, the hash function is only ever used as an argument to the expander.

File: security/s2a/internal/crypter/halfconn.go [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+// UpdateKey updates the traffic secret key, as specified in
+// https://tools.ietf.org/html/rfc8446#section-7.2
+func (hc *S2AHalfConnection) UpdateKey() error {
+	hc.mutex.Lock()
+	defer hc.mutex.Unlock()
+
+	var err error
+	hc.trafficSecret, err = hc.deriveSecret(hc.trafficSecret, []byte(tls13Update), hc.cs.trafficSecretSize())
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v, %v) failed with error: %v", hc.trafficSecret, tls13Update, hc.cs.trafficSecretSize(), err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:95]: Please omit the traffic secret from the logs. In this instance, it would be better to just say e.g. `Failed to advance traffic secret.`

File: security/s2a/internal/crypter/halfconn.go [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+	hc := S2AHalfConnection{cs: cs, h: cs.hashFunction(), expander: &defaultHKDFExpander{}, seqCounter: newCounter(), trafficSecret: trafficSecret}
+
+	var err error
+	if err = hc.updateWithNewTrafficSecret(hc.trafficSecret); err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("hc.updateWithNewTrafficSecret(%v) failed with error: %v", hc.trafficSecret, err)
+	}
+
+	hc.aeadCrypter, err = cs.aeadCrypter(hc.key)
+	if err != nil {
+		return S2AHalfConnection{}, fmt.Errorf("cs.aeadCrypter(%v) failed with error: %v", hc.key, err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:48]: Please don't have the key in the logs. It would be better to say e.g. `Failed to initialize AEAD crypter.`

File: security/s2a/internal/crypter/halfconn.go [hunk 18]
```diff
@@ -0,0 +1,165 @@
...
+	trafficSecret []byte
+	nonce         []byte
+	key           []byte
+}
+
+// NewHalfConn creates a new instance of S2AHalfConnection.
+func NewHalfConn(ciphersuite s2a_proto.Ciphersuite, trafficSecret []byte) (S2AHalfConnection, error) {
+	cs := newCiphersuite(ciphersuite)
+	if cs.trafficSecretSize() != len(trafficSecret) {
+		return S2AHalfConnection{}, fmt.Errorf("supplied traffic secret must be %v bytes, given: %v", cs.trafficSecretSize(), trafficSecret)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:36]: Please do not leak the traffic secret in the logs.

File: security/s2a/internal/crypter/halfconn.go [hunk 19]
```diff
@@ -0,0 +1,165 @@
...
+		hc.mutex.Unlock()
+		return nil, err
+	}
+	nonce := hc.maskedNonce(sequence)
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.decrypt(dst, ciphertext, nonce, aad)
+}
+
+// UpdateKey updates the traffic secret key, as specified in
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:86]: s/updates/advances

You should also mention that it derives the new key and nonce, and resets the sequence number.

File: security/s2a/internal/crypter/halfconn.go [hunk 20]
```diff
@@ -0,0 +1,165 @@
...
+	defer hc.mutex.Unlock()
+
+	var err error
+	hc.trafficSecret, err = hc.deriveSecret(hc.trafficSecret, []byte(tls13Update), hc.cs.trafficSecretSize())
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v, %v) failed with error: %v", hc.trafficSecret, tls13Update, hc.cs.trafficSecretSize(), err)
+	}
+
+	if err = hc.updateWithNewTrafficSecret(hc.trafficSecret); err != nil {
+		return fmt.Errorf("hc.updateWithNewTrafficSecret(%v) failed with error: %v", hc.trafficSecret, err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:99]: Please omit the traffic secret from the log.

File: security/s2a/internal/crypter/halfconn.go [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	err = hc.aeadCrypter.updateKey(hc.key)
+	if err != nil {
+		return fmt.Errorf("hc.aeadCrypter.updateKey(%v) failed with error: %v", hc.key, err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:104]: Please omit the key from the logs.

File: security/s2a/internal/crypter/halfconn.go [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+	return nil
+}
+
+// updateWithNewTrafficSecret takes a new traffic secret and updates the key
+// and nonce.
+func (hc *S2AHalfConnection) updateWithNewTrafficSecret(newTrafficSecret []byte) error {
+	var err error
+	hc.key, err = hc.deriveSecret(newTrafficSecret, []byte(tls13Key), hc.cs.keySize())
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v, %v) failed with error: %v", hc.trafficSecret, tls13Key, hc.cs.keySize(), err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:117]: Please remove the traffic secret from the logs.

File: security/s2a/internal/crypter/halfconn.go [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+	}
+
+	hc.nonce, err = hc.deriveSecret(newTrafficSecret, []byte(tls13Nonce), hc.cs.nonceSize())
+	if err != nil {
+		return fmt.Errorf("hc.deriveSecret(h, %v, %v, %v) failed with error: %v", hc.trafficSecret, tls13Nonce, hc.cs.nonceSize(), err)
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:122]: Please remove the traffic secret from the logs.

File: security/s2a/internal/crypter/halfconn.go [hunk 24]
```diff
@@ -0,0 +1,165 @@
...
+import (
+	"fmt"
+	"golang.org/x/crypto/cryptobyte"
+	s2a_proto "google.golang.org/grpc/security/s2a/internal/proto"
+	"hash"
+	"sync"
+)
+
+// uint64Size is the size of a uint64 in bytes.
+const uint64Size = 8
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:12]: Why is this needed? I thought we removed this from the counter.

File: security/s2a/internal/crypter/halfconn.go [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+	crypter := hc.aeadCrypter
+	hc.mutex.Unlock()
+	return crypter.decrypt(dst, ciphertext, nonce, aad)
+}
+
+// UpdateKey updates the traffic secret key, as specified in
+// https://tools.ietf.org/html/rfc8446#section-7.2
+func (hc *S2AHalfConnection) UpdateKey() error {
+	hc.mutex.Lock()
+	defer hc.mutex.Unlock()
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:90]: This will unlock the mutex whenever the function exits, correct?

File: security/s2a/internal/crypter/halfconn.go [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+func (hc *S2AHalfConnection) getAndIncrementSequence() (uint64, error) {
+	sequence, err := hc.seqCounter.value()
+	if err != nil {
+		return 0, err
+	}
+	hc.seqCounter.increment()
+	return sequence, nil
+}
+
+// maskedNonce creates a new S2A nonce using the sequence number.
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:137]: s/a new S2A nonce using/a copy of the nonce that is masked with/

File: security/s2a/internal/crypter/halfconn.go [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+func (hc *S2AHalfConnection) maskedNonce(sequence uint64) []byte {
+	nonce := make([]byte, len(hc.nonce))
+	copy(nonce, hc.nonce)
+	for i := 0; i < uint64Size; i++ {
+		nonce[nonceSize-uint64Size+i] ^= byte(sequence >> uint64(56-uint64Size*i))
+	}
+	return nonce
+}
+
+// deriveSecret implements Derive-Secret specified in
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:147]: s/implements Derive-Secret/implements the Derive-Secret function, as

File: security/s2a/internal/crypter/halfconn.go [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+// https://tools.ietf.org/html/rfc8446#section-7.1.
+func (hc *S2AHalfConnection) deriveSecret(secret, label []byte, length int) ([]byte, error) {
+	var hkdfLabel cryptobyte.Builder
+	hkdfLabel.AddUint16(uint16(length))
+	hkdfLabel.AddUint8LengthPrefixed(func(b *cryptobyte.Builder) {
+		b.AddBytes(label)
+	})
+	// Append empty `Context` field, specified in the RFC. The Half Connection
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:155]: s/Half Connection/half connection

Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn.go:155]: s/Append empty `Context` field,/Append an empty `Context` field to the label, as

File: security/s2a/internal/crypter/halfconn_test.go [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+		t.Fatalf("Encrypt(%v, %v, nil) failed, err = %v", buf[:0], buf, err)
+	}
+
+	// Encrypt second message.
+	const plaintext2 = "This is a second plaintext."
+	buf2 := []byte(plaintext2)
+	ciphertext2, err := sender.Encrypt(buf2[:0], buf2, nil)
+	if err != nil {
+		t.Fatalf("Encrypt(%v, %v, nil) failed, err = %v", buf2[:0], buf2, err)
+	}
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn_test.go:39]: Please add a third plaintext which is empty.

File: security/s2a/internal/crypter/halfconn_test.go [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+				t.Fatalf("Encrypt(%v, %v, nil) failed, err = %v", buf3[:0], buf3, err)
+			}
+			if got, want := ciphertext3, tc.expectedCiphertext3; !bytes.Equal(got, want) {
+				t.Fatalf("Encrypt(%v, %v, nil) = %v, want %v", buf3[:0], buf3, got, want)
+			}
+		})
+	}
+}
+
+func TestS2AHalfConnectionRoundtrip(t *testing.T) {
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn_test.go:193]: In the test vectors, please change `expectedTrafficSecret` to `advancedTrafficSecret`.

File: security/s2a/internal/crypter/halfconn_test.go [hunk 31]
```diff
@@ -0,0 +1,264 @@
...
+func TestS2AHalfConnectionEncrypt(t *testing.T) {
+	for _, tc := range []struct {
+		ciphersuite                                                  s2a_proto.Ciphersuite
+		trafficSecret                                                []byte
+		expectedCiphertext, expectedCiphertext2, expectedCiphertext3 []byte
+	}{
+		{
+			ciphersuite:         s2a_proto.Ciphersuite_AES_128_GCM_SHA256,
+			trafficSecret:       testutil.Dehex("6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b6b"),
+			expectedCiphertext:  testutil.Dehex("f2e4e411ac674e01385e7ae9db54c97a9ae3d1842e51"),
```
Comment (matthewstevenson88) [inline security/s2a/internal/crypter/halfconn_test.go:132]: Where are the ciphertext values coming from?

File: security/s2a/internal/crypter/common.go (hunk 10 above)
Comment (matthewstevenson88) [reply security/s2a/internal/crypter/common.go]: It doesn't need to be a global constant though. Can we have it be a local constant in the `maskedNonce` function?

**Comment by matthewstevenson88** [comment]
The PR is looking good Ryan! I left a bunch of comments.

One additional test case that might be good: can you set up a test that makes concurrent calls to `Encrypt` in 2 different threads and checks that it does what it should? If this is difficult in Go, we can ignore it but it could be worth asking Cesar about this.
